feat: Add initial AuthzQuerier implementation

Emyrk opened the pull request:
# What this does

This adds the initial AuthzQuerier implementation as an experiment. It has negative latency impacts to the api at this point, but it is correct* (bugs not withstanding).

The goal of this PR is to push a working implementation.

# Testing

Prior to this PR we used `AuthorizeAllEndpoints` as our rbac assertions. This was weak (see https://github.com/coder/coder/issues/5204). There is a similar unit test in `authzquery` that asserts all interface methods have been tested with the proper rbac checks: https://github.com/coder/coder/blob/e6d5c2f4a7102b1a55f374a184a395d5a63d1d48/coderd/authzquery/methods_test.go#L69-L69 It solves the problems brought up in that issue.

So we can remove these tests _in a future PR_:
- https://github.com/coder/coder/blob/e6d5c2f4a7102b1a55f374a184a395d5a63d1d48/coderd/coderdtest/authorize.go#L28-L28
- https://github.com/coder/coder/blob/e6d5c2f4a7102b1a55f374a184a395d5a63d1d48/coderd/coderdtest/authorize_test.go#L10-L10
- https://github.com/coder/coder/blob/e6d5c2f4a7102b1a55f374a184a395d5a63d1d48/enterprise/coderd/coderdenttest/coderdenttest_test.go#L24-L24

Fixes https://github.com/coder/coder/issues/5204

# Future Work

In no particular order

- Implement authz call caching (optimization)
  - Then load testing
- Optimize certain routes (job cancel)
- Use joins to make more tables first class rbac objects (dependant on certain tooling changes)
- ~~Refactor AuthzStore to not be 1:1 with `db.Store`. Make a new interface for it.~~
- Fully implement scoped system roles
- https://github.com/coder/coder/blob/4e6b43f092020dd63a66e6696150cb70fdabb58d/coderd/database/modelmethods.go#L104-L106

## Optimizations

- Do [`workspaceData`](https://github.com/coder/coder/blob/432a261b5a337a436219dfcd8c4909ec5eb3ea4c/coderd/workspaces.go#L1047-L1047) in this layer.

Emyrk committed:
Fix typo

johnstcn committed:
Merge remote-tracking branch 'origin/main' into authzquerier_layer

johnstcn committed:
make RecordingAuthorizer wrap another rbac.Authorizer

johnstcn committed:
fix FakeAuthorizer

johnstcn committed:
skip TestAuthorizeAllEndpoints if authz_querier experiment is enabled

johnstcn committed:
lock more things

johnstcn committed:
Merge remote-tracking branch 'origin/main' into authzquerier_layer

johnstcn commented on coderd/rbac/builtin.go:
```diff
@@ -17,6 +17,12 @@ const (
 	orgAdmin  string = "organization-admin"
 	orgMember string = "organization-member"
+
+	// The below roles are for system internal use only and are
+	// not assignable to users.
+	system         string = "system"
+	systemReadOnly string = "system-read-only"
+	autostart      string = "auto-start"
```
Remove constants, and have the role names only returnable hidden in the function call.

johnstcn commented on coderd/rbac/builtin.go:
```diff
@@ -203,6 +242,12 @@ var (
 	// The first key is the actor role, the second is the roles they can assign.
 	//	map[actor_role][assign_role]<can_assign>
 	assignRoles = map[string]map[string]bool{
+		system: {
```
Add a comment that system is explicitly NOT included in the list of built-in roles so that it can never be assigned to a user.

johnstcn commented on coderd/authzquery/authz_test.go:
```diff
@@ -0,0 +1,206 @@
+			})
+			require.NoError(t, err, "insert workspace")
+
+			data[name] = workspace
+		case database.WorkspaceBuild:
+			build, err := db.InsertWorkspaceBuild(ctx, database.InsertWorkspaceBuildParams{
+				ID:                testCase.Lookup(name),
+				CreatedAt:         time.Now(),
+				UpdatedAt:         time.Now(),
+				WorkspaceID:       takeFirst(orig.WorkspaceID, uuid.New()),
+				TemplateVersionID: takeFirst(orig.TemplateVersionID, uuid.New()),
+				BuildNumber:       takeFirst(orig.BuildNumber, 0),
+				Transition:        takeFirst(orig.Transition, database.WorkspaceTransitionStart),
+				InitiatorID:       takeFirst(orig.InitiatorID, uuid.New()),
+				JobID:             takeFirst(orig.InitiatorID, uuid.New()),
+				ProvisionerState:  []byte{},
+				Deadline:          time.Now(),
+				Reason:            takeFirst(orig.Reason, database.BuildReasonInitiator),
+			})
+			require.NoError(t, err, "insert workspace build")
+
+			data[name] = build
+		case database.User:
+			user, err := db.InsertUser(ctx, database.InsertUserParams{
+				ID:             testCase.Lookup(name),
+				Email:          takeFirst(orig.Email, namesgenerator.GetRandomName(1)),
+				Username:       takeFirst(orig.Username, namesgenerator.GetRandomName(1)),
+				HashedPassword: []byte{},
+				CreatedAt:      time.Now(),
+				UpdatedAt:      time.Now(),
+				RBACRoles:      []string{},
+				LoginType:      takeFirst(orig.LoginType, database.LoginTypePassword),
+			})
+			require.NoError(t, err, "insert user")
+
+			data[name] = user
+		}
+	}
+	return data
+}
```
Extract to a separate PR

johnstcn committed:
rbac/builtin.go: remove consts

johnstcn committed:
extract getAgentSubject()

johnstcn committed:
Merge remote-tracking branch 'origin/main' into authzquerier_layer

johnstcn committed:
use systemCtx in API.oauthLogin()

johnstcn committed:
workspaceagents: fetch request ctx after httpmw.WorkspaceAgent sets authz subject

johnstcn committed:
httpmw: pass systemCtx to getAgentSubject, add OwnerID to workspace agent scopes

johnstcn committed:
authzquery: workspace: fix GetWorkspaceAppByAgentIDAndSlug and GetWorkspaceAppsByAgentID

johnstcn committed:
steven said its ok to remove this

Emyrk committed:
Fix recursive test

Emyrk committed:
Move experiment init below authz init

johnstcn committed:
add httpmw.SystemAuthCtx to api.handleSubdomainApplications

johnstcn committed:
REVERT THIS COMMIT BEFORE MERGING !!!!

johnstcn committed:
ALSO DO NOT MERGE THIS COMMIT

johnstcn committed:
authzquery: fix InsertAgentStat

johnstcn committed:
activitybump: use systemCtx for activityBumpWorkspace

Emyrk committed:
Merge remote-tracking branch 'origin/main' into authzquerier_layer

Emyrk committed:
remove unused function

johnstcn committed:
authzquery: fixes to templates and parameters

 - add doc comment to authorizedQueryWithRelated
 - handle sql.ErrNoRows in parameterRBACResource()
 - fix incorrect logic in GetTemplateVersionByOrganizationAndName

Emyrk committed:
Fix fetch dry run template version from job id

We need to find a better solution imo

Emyrk committed:
Pass actor to follow logs for subscriber listen

Also fix some dry run resource fetching in authzquerier

johnstcn committed:
gerge remote-tracking branch 'origin/main' into authzquerier_layer

johnstcn committed:
rbac: add IsUnauthorizedError, return 404 if UnauthorizedError in organizationByUserAndName

johnstcn committed:
goimports

Emyrk committed:
Implemented first draft testing framework

johnstcn committed:
authzquery: fixes in workspaces.go

- GetWorkspaceAgentsByResourceIDs: handle workspace agents created by
  TemplateVersionImport jobs
- GetWorkspaceResourcesByJobID: handle all provisioner job
  types and simplify RBAC logic

johnstcn committed:
Merge remote-tracking branch 'origin/authzquerier_layer' into authzquerier_layer

Emyrk committed:
Add test method accounting to ensure all functions are called

johnstcn committed:
fixup! authzquery: fixes in workspaces.go

johnstcn committed:
Merge remote-tracking branch 'origin/authzquerier_layer' into authzquerier_layer

Emyrk committed:
Add rbac checks

Emyrk committed:
Fix scim unit tests

johnstcn committed:
authzquery: update UpdateTemplateDeletedByID to call SoftDeleteTemplateByID

johnstcn committed:
Merge remote-tracking branch 'origin/authzquerier_layer' into authzquerier_layer

Emyrk committed:
Fix scim and workspace agent unit tests

Emyrk committed:
Fix getTemplateVersionsByID

Emyrk committed:
Fix more unit tests

Emyrk committed:
Fix license unit test

johnstcn committed:
authzquery: add some more convenience methods, comments etc.

johnstcn committed:
Merge remote-tracking branch 'origin/authzquerier_layer' into authzquerier_layer

Emyrk commented on coderd/authzquery/workspace.go:
```diff
@@ -209,9 +209,9 @@ func (q *AuthzQuerier) GetWorkspaceByID(ctx context.Context, id uuid.UUID) (data
+// Name    string    `db:"name" json:"name"`
 
 // GetWorkspaceByOwnerIDAndName
 // Gen: Workspace
```
We can delete these comments. Whoops

Emyrk committed:
Add sentinel errors for unauth authz errors

Emyrk committed:
Use sentinal error that returns a 404

Emyrk committed:
Use sentinel error always

johnstcn committed:
add slice.New util function

johnstcn committed:
RecordingAuthorizer: AllAsserted: provide more information on missed calls

johnstcn committed:
Merge remote-tracking branch 'origin/authzquerier_layer' into authzquerier_layer

johnstcn committed:
skip GetAuthorizedWorkspaces

Emyrk committed:
Add admin context to provisonerd

Emyrk committed:
Merge remote-tracking branch 'origin/main' into authzquerier_layer

Emyrk committed:
Fix Delte group

Emyrk committed:
remove excess comments

Emyrk committed:
typos and lint

Emyrk committed:
Fix template admin permissions

johnstcn committed:
fix GetLatestWorkspaceBuildsByWorkspaceIDs

Emyrk committed:
Fix rbac unit test

Emyrk committed:
Call compileToSQL in getWorkspaces

Emyrk committed:
Call compileToSQL in getWorkspaces

Emyrk committed:
Fix compile issue

Emyrk committed:
Handle nil prepared case

Emyrk committed:
Linting

johnstcn committed:
add existing workspace tests

Emyrk committed:
Check returned error from db call

Emyrk committed:
Fix build number to be 1 indexed

johnstcn committed:
more tests

johnstcn committed:
generate random AuthInstanceID, more unit tests

Emyrk committed:
Test all api key methods

Emyrk committed:
Test audit methods

Emyrk committed:
Add group and file unit tests

Emyrk committed:
Add template unit test

Emyrk committed:
Add system functions

Emyrk committed:
Merge remote-tracking branch 'origin/main' into authzquerier_layer

Emyrk committed:
Fix merge compile issues

Emyrk committed:
Jobs, orgs, and extra methods implemented

Emyrk committed:
Merge remote-tracking branch 'origin/main' into authzquerier_layer

Emyrk committed:
:

Emyrk committed:
Implement parameters tests

Emyrk committed:
Start license unit tests

Emyrk committed:
Merge remote-tracking branch 'origin/main' into authzquerier_layer

Emyrk committed:
Finish license tests

Emyrk committed:
Add workspace tests

Emyrk committed:
chore: Add WorkspaceApps to dbgen

Emyrk committed:
Add user unit tests

Emyrk committed:
GitSSHKey, UserLink, GitAuthLink

Emyrk committed:
Fix user unit tests

spikecurtis commented on coderd/authzquery/template.go:
```diff
@@ -0,0 +1,319 @@
+		if err != nil {
+			return database.TemplateVersion{}, err
+		}
+	}
```
it would feel more consistent if you encapsulated the logic of querying the DB for the template or using `rbac.ResourceTemplate.InOrg(arg.OrganizationID)` into a `fetchRelated` function, then called `authorizedQueryWithRelated`

Emyrk replied:
Fair. We were on the fence if manual, or using the generic functions was easier to read in many cases

spikecurtis commented on coderd/authzquery/apikey.go:
```diff
@@ -0,0 +1,40 @@
+)
+
+func (q *AuthzQuerier) DeleteAPIKeyByID(ctx context.Context, id string) error {
+	return authorizedDelete(q.logger, q.authorizer, q.database.GetAPIKeyByID, q.database.DeleteAPIKeyByID)(ctx, id)
```
nit: these lines should be broken at something reasonable like 100 chars

Emyrk replied:
Oof, my IDE line is set to 120. We do go over this often too though.

spikecurtis commented on coderd/authzquery/authz.go:
```diff
@@ -0,0 +1,314 @@
+
+// authorizedFetchSet is like authorizedFetch, but works with lists of objects.
+// SQL filters are much more optimal.
+func authorizedFetchSet[ArgumentType any, ObjectType rbac.Objecter,
```
```suggestion
func fetchWithPostFilter[ArgumentType any, ObjectType rbac.Objecter,
```

spikecurtis replied:
I was confused if this was set like "set theory" or like "setting".  It's neither!

spikecurtis commented on coderd/authzquery/authz.go:
```diff
@@ -0,0 +1,314 @@
+// authorizedFetchAndExecWithConverter uses authorizedFetchAndQueryWithConverter but
+// only cares about the error return type. SQL execs only return an error.
+// See authorizedFetchAndQueryWithConverter for more details.
+func authorizedFetchAndExec[ObjectType rbac.Objecter,
```
having everything start with `authorized` feels tedious and doesn't match typical Golang style of omitting contextual prefixes

spikecurtis commented on coderd/authzquery/authz.go:
```diff
@@ -0,0 +1,314 @@
+	// Arguments
+	logger slog.Logger,
+	authorizer rbac.Authorizer,
+	action rbac.Action,
```
I wonder if it is really a good idea to parameterize the action here.  In all sensible cases, the action _should_ be `ActionCreate` and if it is not, it is a code smell and maybe we shouldn't be creating shortcuts for that sort of thing?

Emyrk replied:
You are right, there is like 1 or two edge cases. We can probably handle these manually.

Emyrk replied:
Fixed, no aciton is passed. `ActionCreate` is used.

spikecurtis reviewed (commented):
I like where this is going; some high level comments mostly on the authzquery structure

johnstcn commented on coderd/authzquery/interface.go:
```diff
@@ -0,0 +1,11 @@
+
+// AuthzStore is the interface for the Authz querier. It will track closely
+// to database.Store, but not 1:1 as not all database.Store functions will be
+// exposed.
```
I think one example could be `GetAuthorizedWorkspaces` as `GetWorkspaces` will end up doing the same thing. However, so far the methods have been 1:1.

Emyrk replied:
We were thinking about `System` functions that inherently should never be called by a user being a different interface. That or we need to add some "system" resource in RBAC.

johnstcn commented on coderd/authzquery/authz.go:
```diff
@@ -0,0 +1,314 @@
+	}
+}
+
+func authorizedFetchAndQuery[ObjectType rbac.Objecter, ArgumentType any,
```
Agreed, I also find it confusing. There's definitely room for renaming and refactoring. We might also be able to collapse the number of methods somewhat.

Emyrk replied:
@spikecurtis agree, the generics kinda grew out of necessity and the naming didn't stay up to date. I will spend some time today to clean this up.

I want to replace all `fetch, update, insert` logic with `Exec` and `Query`.

johnstcn committed:
Merge remote-tracking branch 'origin/main' into authzquerier_layer

johnstcn committed:
rm unused-import

johnstcn committed:
authzquery: implement group and system methods

- Use new dbgen methods for tests in system.go
- Implement panicky methods in group.go
- nit: rename Metadatums to Metadata 8-)

johnstcn committed:
fixup! authzquery: implement group and system methods

johnstcn committed:
fixup! authzquery: implement group and system methods

johnstcn committed:
ineffasign

johnstcn committed:
unshadow, unused-reciever

johnstcn committed:
unused-param

johnstcn committed:
finish testing template methods

Emyrk committed:
Rename logger-> log, database->db, authorizer->auth, remove "authorized" prefix

Emyrk commented on coderd/authzquery/authz.go:
```diff
@@ -0,0 +1,314 @@
+	// Arguments
+	logger slog.Logger,
+	authorizer rbac.Authorizer,
+	action rbac.Action,
```
Yes you are right

Emyrk replied:
Remove it, it is just `fetch`

Emyrk committed:
Rename fetchSet to fetchWithPostFilter

Emyrk committed:
Verify the correct error is returned on disallow auth

Emyrk committed:
Linting

johnstcn committed:
database: add missing argument to GetAuthorizedWorkspaces

Emyrk committed:
Refactor recording authorizer

johnstcn commented on coderd/coderdtest/authorize.go:
```diff
@@ -539,52 +542,187 @@ func (a *AuthTester) Test(ctx context.Context, assertRoute map[string]RouteCheck
 }
 
-func (r *RecordingAuthorizer) reset() {
+func (r *RecordingAuthorizer) Reset() {
```
TODO: front-load this upstream

Emyrk committed:
Address incorrect errors

Emyrk committed:
Support asserting outputs in authzquery test

Emyrk committed:
Require outputs to be asserted

Emyrk committed:
Fix comment

Emyrk committed:
allow skipping outputs

Emyrk committed:
Fix user tests to expect outputs

Emyrk committed:
fix api key unit tests to expect outputs

johnstcn committed:
values audit_test.go

Emyrk committed:
Implement outputs for workspace tests

Emyrk committed:
Some system outputs

johnstcn committed:
values file_test.go

johnstcn committed:
values group_test.go

Emyrk committed:
Template outputs

Emyrk committed:
System outputs

johnstcn committed:
values job_test.go, methods_test.go

Emyrk committed:
Add organization output

johnstcn committed:
values license_test.go

johnstcn committed:
Merge remote-tracking branch 'origin/authzquerier_layer' into authzquerier_layer

Emyrk committed:
Add parameters ooutput

Emyrk committed:
Api key and audit fix

Emyrk committed:
Fix file outputs

Emyrk committed:
Fix groups

Emyrk committed:
Fix job, license, and org

Emyrk committed:
System done

Emyrk committed:
Fix templates

Emyrk committed:
Fix most users

Emyrk committed:
Linting

johnstcn committed:
workspace_test.go values fix

johnstcn committed:
nolint unreachable

Emyrk committed:
Fix all user method tests

Emyrk committed:
Add unit tests for InTx and Ping

johnstcn commented on coderd/coderdtest/coderdtest.go:
```diff
@@ -179,12 +181,13 @@ func NewOptions(t *testing.T, options *Options) (func(http.Handler), context.Can
-		// 	options.Authorizer = &RecordingAuthorizer{}
-		// }
-		// options.Database = authzquery.NewAuthzQuerier(options.Database, options.Authorizer)
+	if strings.Contains(os.Getenv("CODER_EXPERIMENTS_TEST"), "authz_querier") || true {
```
@johnstcn DO NOT forget to remove the `|| true` here

Emyrk committed:
Add AuthorizedXX tests

johnstcn commented on coderd/coderdtest/authorize_test.go:
```diff
@@ -9,10 +9,14 @@ import (
 func TestAuthorizeAllEndpoints(t *testing.T) {
 	t.Parallel()
+	// TODO: DO NOT MERGE THIS
+	t.Skip("TODO: fix all the unit tests that break when this is enabled. ")
```
@johnstcn :eyes:

johnstcn commented on enterprise/coderd/coderdenttest/coderdenttest_test.go:
```diff
@@ -23,6 +23,8 @@ func TestNew(t *testing.T) {
 func TestAuthorizeAllEndpoints(t *testing.T) {
 	t.Parallel()
+	// TODO: DO NOT MERGE THIS
+	t.Skip("TODO: fix all the unit tests that break when this is enabled. ")
```
@johnstcn 👀

johnstcn committed:
api: skip Authorize if codersdk.ExperimentAuthzQuerier enabled

johnstcn committed:
Merge remote-tracking branch 'origin/authzquerier_layer' into authzquerier_layer

Emyrk committed:
Only abort early on checks that should be removed

Emyrk commented on coderd/authzquery/authzquerier.go:
```diff
@@ -0,0 +1,67 @@
+// Use WithAuthorizeContext to set the authorization subject in the context for
+// the common user case.
+type AuthzQuerier struct {
+	database   database.Store
```
@spikecurtis once this is in, we want to have a new interface for this object. So no where in the code will have access to a `db.Store`.

Emyrk replied:
Actually this needs to be a conversation. Will be a follow up PR to enforce this protection

Emyrk committed:
remove authorizedQuery

johnstcn committed:
authzquery: use GetProvisionerJobById to auth GetWorkspaceResourceByID

johnstcn committed:
Merge remote-tracking branch 'origin/authzquerier_layer' into authzquerier_layer

Emyrk committed:
All insert generic functions use rbac.ActionCreate

Emyrk committed:
Fix unit tests that use create over update

johnstcn committed:
un-skip TestAuthorizeAllEndpoints and remove always-true conditional for authzquerier unit tests

johnstcn committed:
Merge remote-tracking branch 'origin/authzquerier_layer' into authzquerier_layer

johnstcn committed:
Merge remote-tracking branch 'origin/main' into authzquerier_layer

johnstcn committed:
fixup! un-skip TestAuthorizeAllEndpoints and remove always-true conditional for authzquerier unit tests

johnstcn committed:
where my members at yo

Emyrk committed:
Allow out of order slicing

Emyrk committed:
Use slice.New()

johnstcn committed:
paralalalaleleleel

Emyrk committed:
Ordering of users in fetch

Emyrk committed:
Add actual scope to workspace agent ctx

Emyrk committed:
Merge remote-tracking branch 'origin/main' into authzquerier_layer

spikecurtis commented on coderd/database/modelmethods.go:
```diff
@@ -121,8 +134,26 @@ func (u User) UserDataRBACObject() rbac.Object {
+
+func (u GitAuthLink) RBACObject() rbac.Object {
+	// I assume UserData is ok?
+	return rbac.ResourceUserData.WithID(u.UserID).WithOwner(u.UserID.String())
```
this looks like the only actual usage of `ResourceUserData` -- why did we add it in the first place?

spikecurtis replied:
I did a "find usage" on `ResourceUserData` on `main` and the only reference is tests.  So, I don't think the current implementation actually checks for `ResourceUserData` before doing things like updating user passwords/profiles/GitSSHKeys.

Emyrk replied:
This is from `main`. https://github.com/coder/coder/blob/33f2c8fef5fed25209d08896ca20e01d1655ea91/coderd/gitsshkey.go#L38-L41

It was supposed to, but the modelmethod was incorrect.

spikecurtis commented on coderd/database/modelmethods.go:
```diff
@@ -121,8 +134,26 @@ func (u User) UserDataRBACObject() rbac.Object {
+}
+
+func (l License) RBACObject() rbac.Object {
+	return rbac.ResourceLicense.WithIDString(strconv.FormatInt(int64(l.ID), 10))
```
are we using the `ID` or `UUID`?  I'm not clear what's going on with licenses in this regard, and it's an example of why we should only have one identifier.

Emyrk replied:
Honestly I was not 100% sure, but not all Licenses had a `uuid`. It is a nullable field. I was adding the `ID` in the rbac mainly for logging reasons since the the RBAC input is logged.

spikecurtis commented on coderd/database/modelqueries.go:
```diff
@@ -200,6 +200,7 @@ func (q *sqlQuerier) GetAuthorizedWorkspaces(ctx context.Context, arg GetWorkspa
 	rows, err := q.db.QueryContext(ctx, query,
 		arg.Deleted,
 		arg.Status,
+		pq.Array(arg.WorkspaceIds),
```
🤨

johnstcn replied:
We added WorkspaceIDs to the GetAuthorizedWorkspaces query to replace `GetWorkspacesWithIDs`.

Emyrk replied:
will revert the change, it is no longer used.

Emyrk commented on coderd/authzquery/system.go:
```diff
@@ -0,0 +1,194 @@
+// TODO: @emyrk should we name system functions differently to indicate a user
+// cannot call them? Maybe we should have a separate interface for system functions?
+// So you'd do `authzQ.System().GetDERPMeshKey(ctx)` or something like that?
+// Cian: yes. Let's do it.
```
:+1: We can make all these things rbac resources then.

Emyrk replied:
Fixed comment, will add to Future work.

spikecurtis replied:
To be clear, my position is that these 

a) need an authorization context
b) need an authorization check based on the context

It _may_ not be necessary for the check in b) to be an _RBAC check_.  For example, we could just say that these methods are authorized if and only if the authorization context is the provisioner context.  Calling OPA is expensive and if we we know that user contexts are never allowed to call these methods, it's simpler and faster to just directly check the principal.

Emyrk replied:
:+1: This is good feedback.

Emyrk commented on coderd/database/modelmethods.go:
```diff
@@ -94,6 +100,13 @@ func (m OrganizationMember) RBACObject() rbac.Object {
 }
 
+func (m GetOrganizationIDsByMemberIDsRow) RBACObject() rbac.Object {
+	// TODO: This feels incorrect as we are really returning a list of orgmembers.
```
Yes, this maintains what we checked before. The calling code still uses the returned `User` fields, so I was not able to change this to an `OrganizationMember` response.

It requires some refactoring in the route code, so I did not change it in this PR.

spikecurtis replied:
The `rbac.Object` for this is independent of the returned type (`GetOrganizationIDsByMemberIDsRow`) so we can fix the RBAC check without refactoring the route.

Emyrk replied:
Correct. I can fix this. My original objection to correcting it was that the returned data was `User`. So I was matching the payload.

I will fix the RBAC check to `ResourceOrganizationMember`, and will change the payload in a future PR.

Emyrk committed:
RBAC UserData should use the correct rbac resource

Emyrk committed:
Remove workspace IDs filter arg

johnstcn committed:
rename authzquery.NewAuthzQuerier to authzquery.New

Emyrk committed:
Start removing QueryByRelated

Emyrk committed:
Start removing QueryByRelated

johnstcn committed:
remove queryWithRelated

Emyrk committed:
Fixup generic func comments

johnstcn committed:
fixup! remove queryWithRelated

Emyrk committed:
remove todo

Emyrk committed:
Improve readability of generics and arguments

Emyrk commented on coderd/authzquery/authz.go:
```diff
@@ -0,0 +1,299 @@
+		})(ctx, arg)
+		return err
+	}
+}
```
Fixed

Emyrk commented on enterprise/coderd/coderd.go:
```diff
@@ -144,7 +144,11 @@ func New(ctx context.Context, options *Options) (*API, error) {
-			r.Use(api.scimEnabledMW)
+			r.Use(
+				api.scimEnabledMW,
+				// TODO: Make a scim auth role.
```
We do not, we just have it listed in "Future Work". It is easy to find, it's any place `WithAuthorizeSystemContext` is called.

Emyrk committed:
Update fetchAndQuery comment

Emyrk commented on coderd/authzquery/authz.go:
```diff
@@ -0,0 +1,299 @@
+	authorizer rbac.Authorizer,
+	fetchFunc Fetch,
+	updateExec Exec) Exec {
+	return fetchAndExec(logger, authorizer, rbac.ActionUpdate, fetchFunc, updateExec)
```
:+1: We removed `insertWithReturn`, and kept `fetchAndExec`. So we use 1 pattern now.

Emyrk committed:
Fix comment about system functions

johnstcn committed:
remove insert() function

johnstcn committed:
insertWithReturn is the new insert

johnstcn committed:
Merge remote-tracking branch 'origin/authzquerier_layer' into authzquerier_layer

Emyrk committed:
Remove duplicate workspace agent scope

Emyrk committed:
Pass agent ctx into activityBumpWorkspace

Emyrk commented on coderd/activitybump.go:
```diff
@@ -19,6 +21,8 @@ func activityBumpWorkspace(log slog.Logger, db database.Store, workspaceID uuid.
 	// low priority operations fail first.
 	ctx, cancel := context.WithTimeout(context.Background(), time.Second*15)
+	// We always want to use the **system** authz context for this.
+	ctx = authzquery.WithAuthorizeSystemContext(ctx, rbac.RolesAdminSystem())
```
Fixed.

johnstcn committed:
remove panic

johnstcn committed:
Merge remote-tracking branch 'origin/authzquerier_layer' into authzquerier_layer

Emyrk committed:
Remove uneeded comments

Emyrk committed:
Use 's' for all suite methods

Emyrk committed:
Reduce LoC by using setup and teardown test

Emyrk committed:
Remove nested "RunMethodTest", use new assertions

Emyrk committed:
Start converting tests to the new format

Emyrk committed:
refactor out error test

Emyrk committed:
Update unit test teardown to include NoActorError

Emyrk committed:
Attempt a new style of subtest

Emyrk committed:
Fix user tests to use new subtest strategy

Emyrk committed:
Fix unit tests names

Emyrk committed:
Convert more tests to new format

Emyrk committed:
Convert all unit tests

Emyrk committed:
Add comments

Emyrk committed:
remove unused code

Emyrk committed:
rename MethodCase to expects

Emyrk committed:
Merge remote-tracking branch 'origin/main' into authzquerier_layer

Emyrk committed:
DB function was renamed/changed

johnstcn committed:
imports

johnstcn committed:
authzquery -> database/dbauthz

johnstcn committed:
conditionally skip TestAuthorizeAllEndpoints

johnstcn committed:
userauth: use systemCtx when setting user groups

johnstcn committed:
Merge branch 'cj/dbauthz' into authzquerier_layer

johnstcn committed:
fixup! authzquery -> database/dbauthz

johnstcn committed:
rm todo

Emyrk committed:
Condense into 1 file

System.go is still on it's own until a solution is in place

Emyrk committed:
doc.go

kylecarbs commented on coderd/database/dbauthz/authzquerier.go:
```diff
@@ -0,0 +1,103 @@
+package dbauthz
```
This file should be renamed to `dbauthz.go` and the general file in the package.

kylecarbs replied:
Ahh, I see there's already a `dbauthz.go`. Might be cleaner to put this in there... it's odd that the `New` func for the package isn't in the same filename.

Emyrk replied:
Fair. We just need a place to move all those methods too. We merged all the methods into 1 file, just make naming that file a little annoying. I could just name it `methods.go`

Emyrk replied:
Consolidated into `dbauthz.go` and moved the other files to `methods.go`

kylecarbs commented on coderd/database/dbauthz/context.go:
```diff
@@ -0,0 +1,35 @@
+	"github.com/coder/coder/coderd/rbac"
+)
+
+type authContextKey struct{}
```
Could this just be in `dbauthz.go` instead? Seems small and relative to the package as a whole, not a separate part.

kylecarbs commented on coderd/database/dbauthz/crud.go:
```diff
@@ -0,0 +1,228 @@
+// insert runs an rbac.ActionCreate on the rbac object argument before
+// running the insertFunc. The insertFunc is expected to return the object that
+// was inserted.
+func insert[
```
I might toss these functions in `dbauthz.go` too... for the reason that `crud.go` doesn't mean much to me if I'm scanning through files, but I'd certainly expect this in `dbauthz.go`.

kylecarbs commented on coderd/database/dbauthz/doc.go:
```diff
@@ -0,0 +1,17 @@
+// Package dbauthz provides an authorization layer on top of the database. This
```
Wonderful doc 😍

kylecarbs commented on coderd/database/dbauthz/doc.go:
```diff
@@ -0,0 +1,17 @@
+// queries as close to 1 as possible. Each method should do 1 thing, with no
+// unexpected side effects (eg: updating multiple tables in a single method).
+//
+// Avoid implementing business logic in this package. Only authorization related
```
```suggestion
// Do not implement business logic in this package. Only authorization related
```

kylecarbs commented on coderd/database/dbauthz/interface.go:
```diff
@@ -0,0 +1,11 @@
+// AuthzStore is the interface for the Authz querier. It will track closely
+// to database.Store, but not 1:1 as not all database.Store functions will be
+// exposed.
+type AuthzStore interface {
```
Because this package is called `dbauthz`, we can just name this `Store`. Then it fits nicely with `database.Store` too!

I'd toss this file in `dbauthz.go` too. It's weird for the primary interface to be in a separate file.

kylecarbs commented on coderd/database/dbauthz/methods_test.go:
```diff
@@ -0,0 +1,367 @@
+package dbauthz_test
```
Thoughts on calling this `helper_test.go` instead? Seems more like helpers and setup rather than `methods`. Or `setup_test.go` would work too!

kylecarbs commented on coderd/database/dbauthz/system.go:
```diff
@@ -0,0 +1,194 @@
+	"github.com/coder/coder/coderd/database"
+)
+
+// TODO: All these system functions should have rbac objects created to allow
```
Just so I understand, all of these are temporary while we don't have objects associated, right?

Emyrk replied:
The current pass through implementation is temporary yes. We want to follow this up in another PR.

Current solutions include ideas such as `db.Sudo()`, creating new rbac resources, or not using rego authz at all.

----

But until we implement that solution, we kept them in their own file `system.go` so we can easily identify which methods need to be addressed. Remember that this keeps parity with `main` since there are currently no rbac checks for these methods in `main`.

kylecarbs commented on coderd/httpmw/workspaceagent.go:
```diff
@@ -30,6 +32,7 @@ func ExtractWorkspaceAgent(db database.Store) func(http.Handler) http.Handler {
 	return func(next http.Handler) http.Handler {
 		return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
 			ctx := r.Context()
+			systemCtx := dbauthz.WithAuthorizeSystemContext(ctx, rbac.RolesAdminSystem())
```
Hmm, I thought we were getting rid of this?

johnstcn replied:
I discussed this with @spikecurtis this morning and he made the point that keeping the actor available in the context even for the system user will be useful down the line for audit logging purposes. The actual change itself is comparatively and can be made in a separate PR if required.

kylecarbs replied:
I will say that this doesn't make a lot of sense to me as a reader of this code... gaining a privileged context that when passed in to a database function bypasses the authorization is *very* abstract, and really hard to track down what's actually going on.

I'd rather we just pass a `privilegedDB` object around that references the PG database directly without the authz wrapper, this just seems really confusing to me.

spikecurtis replied:
I'm very strongly against passing the underlying `database.Store` object around.

The way to think about this is that

a) we always have a "principal" for every action
b) we always do an authorization check

So, the privileged context doesn't _turn off or bypass_ authentication, it changes the principal to a privileged one, and that principal always _passes_ authentication checks.  This maintains the same set of abstractions as other callers to the `Store`.

It is important to ensure that we always have a principal to attribute the action to, because the next security project after we get authz working is to also push audit logging down into the persistence layer.  Dropping audit logs in API handlers suffers the exact same problems as doing authz checks there, and audit logs that are not 100% complete are much less useful.

kylecarbs replied:
From my understanding, a `system` user isn't an actor at all, it's fair to say that it holds a similar level of abstract context-dropping behavior as executing the database directly. If the logs showed me that the `system` deleted a workspace, the only information I'd meaningfully retrieve is the timestamp it occurred at, but nothing else is of value there.

I believe the connected user should authorize all actions and that an initiator or actor could be the system. To me, authorizing through the system is just an abstract way to access the underlying database, which, even if audited, I'm not sure what it provides the user apart from the awareness that _something_ did _something_ at X time.

Emyrk replied:
The challenging thing about this PR is we **have to use ctx** to indicate "system user/action" because it is a `db.Store`. 

The alternate idea of `.Sudo()` returning a privileged db requires addeding `.Sudo()` to `database.Store` which doesn't make much sense? 

In that if we do the `.Sudo()` idea, we probably want 2 different interfaces. One being an `AuthzStore` which is a superset of `db.Store`. Or something like this.

```
type AuthzStore interface{
  database.Store
  Sudo() database.Store
}
```

------

Until this is the way we do things, it would be a bit awkward to have code call `Sudo` on the underlying database type when the underlying type is just a database. We'd probably need to make a `type NoopAuthzQuerier struct`??

Emyrk committed:
Update coderd/database/dbauthz/doc.go

Co-authored-by: Kyle Carberry <kyle@coder.com>

Emyrk committed:
Move files around, consolidate to dbauthz.go

Emyrk committed:
Merge remote-tracking branch 'origin/main' into authzquerier_layer

Emyrk committed:
fix unit test to work with dbauthz

Emyrk commented on coderd/database/dbauthz/dbauthz.go:
```diff
@@ -0,0 +1,323 @@
+//
+// Use WithAuthorizeContext to set the authorization subject in the context for
+// the common user case.
+type AuthzQuerier struct {
+	db   database.Store
+	auth rbac.Authorizer
+	log  slog.Logger
+}
```
Should we name this something else since the package name is dbauthz? Like `Store`? Or just `Querier`?

kylecarbs replied:
Does this need to be exposed? Does the interface not suffice?

johnstcn replied:
We need the tests in a separate package to avoid an import cycle.
Also, our tests check all the methods on `AuthzQuerier` (which will not always be 1:1 with `database.Store`).

Emyrk replied:
@johnstcn we can still get the methods on an unexported struct. We just need to deref the pointer in reflect. (99% sure)

kylecarbs replied:
Meh, it's fine to expose this IMO, was just curious.

I think the name should be `Querier` though, since the package is called `dbauthz`.

Emyrk committed:
Consolidate files

kylecarbs commented on coderd/database/dbauthz/methods.go:
```diff
@@ -0,0 +1,1622 @@
+package dbauthz
```
@Emyrk thoughts on calling this `querier.go` instead of `methods`? Methods feels very Java-esque to me.

johnstcn replied:
Calling them "methods" is [technically correct](https://go.dev/ref/spec#Method_declarations), but I have no preference here.

Emyrk replied:
Yea, I was going with the golang naming of methods being functions attached to a struct. In this case, the primary structure of the package.

johnstcn committed:
goimports

johnstcn committed:
rename methods.go -> querier.go

Emyrk committed:
Do not export the authzQuerier

Emyrk committed:
Rename to "querier", add unit test for double wrap protection

Emyrk commented on coderd/database/dbauthz/querier_test.go:
```diff
@@ -0,0 +1,1206 @@
+			CreatedBy: f.CreatedBy,
+		}).Asserts(f, rbac.ActionRead).Returns(f)
+	}))
+	s.Run("GetFileByID", s.Subtest(func(db database.Store, check *expects) {
```
Unfortunately you cannot. The method signature **must** be `s.Run(name string, test fun())` with no arguments. I tried overriding `Run` with `s.Run(name string, test fun(db database.Store, check *expects))` and other tricks. All my tricks broke the IDE detection of the subtests. 

So although cli `go test -run XXX/YYY` would still work, you don't get the nice green arrow on your IDE :cry:. 

So the `s.Subtest(...)` was my way of returning a function with 0 arguments, but still capturing the `db` and `expects` in the function body nicely.

Emyrk commented on coderd/database/dbauthz/dbauthz.go:
```diff
@@ -0,0 +1,337 @@
+func insert[
+	ObjectType any,
+	ArgumentType any,
+	Insert func(ctx context.Context, arg ArgumentType) (ObjectType, error),
```
Right?!

johnstcn committed:
remove duplicate dbauthz init

johnstcn committed:
use codersdk experiment value instead of hard-coded string

johnstcn committed:
Merge remote-tracking branch 'origin/main' into authzquerier_layer

johnstcn committed:
Merge remote-tracking branch 'origin/main' into authzquerier_layer

Emyrk committed:
Remove rbac ctx from provisionerd

Emyrk committed:
fixup! Remove rbac ctx from provisionerd

johnstcn committed:
wip: dbauthz.WithAuthorizeSystemContext -> dbauthz.AsSystem()

Emyrk committed:
Add lint rule to prevent system ctx abuse

johnstcn committed:
fixup! wip: dbauthz.WithAuthorizeSystemContext -> dbauthz.AsSystem()

johnstcn committed:
Merge remote-tracking branch 'origin/authzquerier_layer' into authzquerier_layer

johnstcn committed:
fix autobuild/executor unit tests

Emyrk committed:
Add middleware for using system ctx in middlewares

johnstcn committed:
fix compile errors

johnstcn committed:
set system ctx in provisionerdserver

Emyrk committed:
Unit test the AsAuthzSystem mw

Emyrk committed:
Update unit tests to cover the no actor case

Emyrk committed:
Typo

Emyrk committed:
remove todo

Emyrk committed:
User proper rbac errors in unit test

Emyrk committed:
Add unit test to cover prepareSQL error case

Emyrk committed:
NullUUID is empty, so takeFirst fails

Emyrk committed:
Add AsSystem

Emyrk committed:
Fix internal error logging

Emyrk committed:
Remove error noise in unit tests

Emyrk committed:
Use AsSystem for decrypting encrypted api keys

johnstcn committed:
fix linter errors

johnstcn committed:
userauth: create API key as user instead of as system

johnstcn committed:
Merge remote-tracking branch 'origin/main' into authzquerier_layer

Emyrk committed:
Remove unused file

Emyrk committed:
Use system context to set a disconnected agent

Emyrk committed:
Log error on failed agent disconnect update

Emyrk committed:
Merge remote-tracking branch 'origin/main' into authzquerier_layer

Emyrk committed:
Unit tests do not handle error log well

Emyrk committed:
Merge remote-tracking branch 'origin/main' into authzquerier_layer

Emyrk committed:
Fix license uuid in merge

Emyrk committed:
Fix unit test error logging

Emyrk committed:
Correct the returned error from not authorized

Emyrk committed:
Fix if/else logic

Emyrk committed:
fixup! Fix if/else logic

johnstcn committed:
Merge remote-tracking branch 'origin/main' into authzquerier_layer
